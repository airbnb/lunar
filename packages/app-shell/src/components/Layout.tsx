import React from 'react';
import useStyles, { StyleSheet } from '@airbnb/lunar/lib/hooks/useStyles';

export const styleSheet: StyleSheet = () => ({
  layout: {
    minHeight: '100vh',
  },
});

export type Props = {
  children: NonNullable<React.ReactNode>;
};

function ShellLayout({ children }: Props) {
  const [styles, cx] = useStyles(styleSheet);

  return <div className={cx(styles.layout)}>{children}</div>;
}

export default React.memo(ShellLayout);
