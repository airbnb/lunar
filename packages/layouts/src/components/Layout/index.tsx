import React from 'react';
import withStyles, { WithStylesProps } from '@airbnb/lunar/lib/composers/withStyles';

export type Props = {
  /** The primary main content. */
  children: NonNullable<React.ReactNode>;
  /** Expand main content to full width of viewport. */
  fluid?: boolean;
  /** Remove background color from main content. */
  noBackground?: boolean;
  /** Remove padding from main content. */
  noPadding?: boolean;
};

export type AsideProps = {
  /** The after aside content. */
  after?: React.ReactNode;
  /** The before aside content. */
  before?: React.ReactNode;
};

/** Abstract layout manager that all other layouts extend from. */
export class Layout extends React.Component<Props & AsideProps & WithStylesProps> {
  static defaultProps = {
    fluid: false,
    noBackground: false,
    noPadding: false,
  };

  render() {
    const { cx, after, before, children, fluid, noBackground, noPadding, styles } = this.props;

    return (
      <div className={cx(styles.layout)}>
        {before}

        <main
          role="main"
          className={cx(
            styles.main,
            noBackground && styles.main_noBackground,
            noPadding && styles.main_noPadding,
          )}
        >
          <div className={cx(!fluid && styles.mainContent)}>{children}</div>
        </main>

        {after}
      </div>
    );
  }
}

export default withStyles(({ breakpoints, color, unit }) => ({
  layout: {
    display: 'flex',
    width: '100%',
    minHeight: '100vh',
    justifyContent: 'space-between',
  },

  main: {
    flexGrow: 1,
    padding: unit * 2,
    background: color.accent.bgHover,
  },

  main_noBackground: {
    background: 'transparent',
  },

  main_noPadding: {
    padding: 0,
  },

  mainContent: {
    maxWidth: breakpoints.medium,
  },
}))(Layout);
