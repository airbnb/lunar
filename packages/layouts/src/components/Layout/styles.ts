import { StyleSheet } from '@airbnb/lunar/lib/hooks/useStyles';

const styleSheet: StyleSheet = ({ breakpoints, color, unit }) => ({
  layout: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-between',
  },

  main: {
    flexGrow: 1,
    padding: `${unit * 2}px ${unit * 3}px`,
    background: color.accent.bgHover,
  },

  main_noBackground: {
    background: 'transparent',
  },

  main_noPadding: {
    padding: 0,
  },

  mainContent: {
    maxWidth: breakpoints.medium,
  },
});

export default styleSheet;

export { styleSheet };
