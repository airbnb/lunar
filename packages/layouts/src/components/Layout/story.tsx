import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Aside from '../Aside';
import Layout from '.';

export default {
  title: 'Layouts/Layout',
  parameters: {
    inspectComponents: [Layout],
  },
};

export function standardLayout() {
  return (
    <Layout>
      <LoremIpsum />
    </Layout>
  );
}

standardLayout.story = {
  name: 'Standard layout.',
};

export function withAFixedHeight() {
  return (
    <Layout minHeight={300}>
      <LoremIpsum />
    </Layout>
  );
}

withAFixedHeight.story = {
  name: 'With a custom min-height.',
};

export function withLeftAsideAndNoMainPadding() {
  return (
    <Layout
      noPadding
      before={
        <Aside width={300}>
          <LoremIpsum />
        </Aside>
      }
    >
      <LoremIpsum />
    </Layout>
  );
}

withLeftAsideAndNoMainPadding.story = {
  name: 'With left aside and no main padding.',
};

export function withRightAsideAndNoMainBackgroundColor() {
  return (
    <Layout
      noBackground
      after={
        <Aside width={300}>
          <LoremIpsum />
        </Aside>
      }
    >
      <LoremIpsum />
    </Layout>
  );
}

withRightAsideAndNoMainBackgroundColor.story = {
  name: 'With right aside and no main background color.',
};

export function withBothSides() {
  return (
    <Layout
      before={
        <Aside width={300}>
          <LoremIpsum />
        </Aside>
      }
      after={
        <Aside width={300}>
          <LoremIpsum />
        </Aside>
      }
    >
      <LoremIpsum />
    </Layout>
  );
}

withBothSides.story = {
  name: 'With both sides.',
};

export function withBothAsidesCollapsibleAndMainPadding() {
  return (
    <Layout
      before={
        <Aside before collapsible width={300}>
          <LoremIpsum />
        </Aside>
      }
      after={
        <Aside after collapsible width={300}>
          <LoremIpsum />
        </Aside>
      }
    >
      <LoremIpsum />
    </Layout>
  );
}

withBothAsidesCollapsibleAndMainPadding.story = {
  name: 'With both sides, collapsible.',
};
