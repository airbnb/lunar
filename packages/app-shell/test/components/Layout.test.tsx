import React from 'react';
import { shallow } from 'enzyme';
import Layout from '../../src/components/Layout';

describe('Layout', () => {
  it('renders all toasts', () => {
    const wrapper = shallow(<Layout>Child</Layout>);

    expect(wrapper).toMatchSnapshot();
  });
});
