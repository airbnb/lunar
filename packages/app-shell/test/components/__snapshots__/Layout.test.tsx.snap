// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders all toasts 1`] = `
<div
  className="layout_149btvt"
>
  Child
</div>
`;
